src/fallback/dom_parser_implementation.cpp

```diff
@@ -243,6 +243,93 @@ WARN_UNUSED error_code implementation::minify(const uint8_t *buf, size_t len, ui
   // for fear of aliasing
   return SUCCESS;
 }
+WARN_UNUSED bool implementation::validate_utf8(const char *buf, size_t len) const noexcept { 
+  const uint8_t *data = (const uint8_t *)buf;
+  uint64_t pos = 0;
+  uint64_t next_pos = 0;
+  uint32_t code_point = 0;
+  while (pos < len) {
+
+    // check of the next 8 bytes are ascii.
+    next_pos = pos + 16;
+    if (next_pos <=
```

Comment:
Absolutely not for this PR, but we could avoid this nested if with a "full block / remainder" algorithm: while (pos <= len - 8) { maybe skip 8 bytes; check utf8 } and then if (pos < len) { check utf8 }. It would also remove almost all of the if (next_pos > len) checks inside of here, which has to be a massive win!

Reply:
Point taken.